File: LogisimPlus/LogisimPlus/draw/shapes/FillableCanvasObject.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace draw.shapes
{

	using CanvasObject = draw.model.CanvasObject;
	using AbstractCanvasObject = draw.model.AbstractCanvasObject;
	using logisim.data;
	using AttributeOption = logisim.data.AttributeOption;

	internal abstract class FillableCanvasObject : AbstractCanvasObject
	{
		private AttributeOption paintType;
		private int strokeWidth;
		private Color strokeColor;
		private Color fillColor;

		public FillableCanvasObject()
		{
			paintType = DrawAttr.PAINT_STROKE;
			strokeWidth = 1;
			strokeColor = Color.Black;
			fillColor = Color.White;
		}

		public override bool matches(CanvasObject other)
		{
			if (other is FillableCanvasObject)
			{
				FillableCanvasObject that = (FillableCanvasObject) other;
				bool ret = this.paintType == that.paintType;
				if (ret && this.paintType != DrawAttr.PAINT_FILL)
				{
					ret = ret && this.strokeWidth == that.strokeWidth && this.strokeColor.Equals(that.strokeColor);
				}
				if (ret && this.paintType != DrawAttr.PAINT_STROKE)
				{
					ret = ret && this.fillColor.Equals(that.fillColor);
				}
				return ret;
			}
			else
			{
				return false;
			}
		}

		public override int matchesHashCode()
		{
			int ret = paintType.GetHashCode();
			if (paintType != DrawAttr.PAINT_FILL)
			{
				ret = ret * 31 + strokeWidth;
				ret = ret * 31 + strokeColor.GetHashCode();
			}
			else
			{
				ret = ret * 31 * 31;
			}
			if (paintType != DrawAttr.PAINT_STROKE)
			{
				ret = ret * 31 + fillColor.GetHashCode();
			}
			else
			{
				ret = ret * 31;
			}
			return ret;
		}

		public virtual AttributeOption PaintType
		{
			get
			{
				return paintType;
			}
		}

		public virtual int StrokeWidth
		{
			get
			{
				return strokeWidth;
			}
		}

		public override object getValue(Attribute attr)
		{
			if (attr == DrawAttr.PAINT_TYPE)
			{
				return paintType;
			}
			else if (attr == DrawAttr.STROKE_COLOR)
			{
				return strokeColor;
			}
			else if (attr == DrawAttr.FILL_COLOR)
			{
				return fillColor;
			}
			else if (attr == DrawAttr.STROKE_WIDTH)
			{
				return Convert.ToInt32(strokeWidth);
			}
			else
			{
				return null;
			}
		}

        protected internal override void updateValue(Attribute attr, object value)
		{
			if (attr == DrawAttr.PAINT_TYPE)
			{
				paintType = (AttributeOption) value;
				fireAttributeListChanged();
			}
			else if (attr == DrawAttr.STROKE_COLOR)
			{
				strokeColor = (Color) value;
			}
			else if (attr == DrawAttr.FILL_COLOR)
			{
				fillColor = (Color) value;
			}
			else if (attr == DrawAttr.STROKE_WIDTH)
			{
				strokeWidth = ((int?) value).Value;
			}
		}
	}

}

File: logisim/app/src/main/java/draw/shapes/FillableCanvasObject.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package draw.shapes;

import java.awt.Color;

import draw.model.CanvasObject;
import draw.model.AbstractCanvasObject;
import logisim.data.Attribute;
import logisim.data.AttributeOption;

abstract class FillableCanvasObject extends AbstractCanvasObject {
	private AttributeOption paintType;
	private int strokeWidth;
	private Color strokeColor;
	private Color fillColor;

	public FillableCanvasObject() {
		paintType = DrawAttr.PAINT_STROKE;
		strokeWidth = 1;
		strokeColor = Color.BLACK;
		fillColor = Color.WHITE;
	}

	@Override
	public boolean matches(CanvasObject other) {
		if (!(other instanceof FillableCanvasObject that))
			return false;
		boolean ret = paintType == that.paintType;
		if (ret && paintType != DrawAttr.PAINT_FILL)
			ret = strokeWidth == that.strokeWidth && strokeColor.equals(that.strokeColor);
		if (ret && paintType != DrawAttr.PAINT_STROKE)
			return fillColor.equals(that.fillColor);
		return ret;
	}

	@Override
	public int matchesHashCode() {
		int ret = paintType.hashCode();
		if (paintType != DrawAttr.PAINT_FILL) {
			ret = ret * 31 + strokeWidth;
			ret = ret * 31 + strokeColor.hashCode();
		} else
			ret = ret * 31 * 31;
		return ret * 31 + (paintType != DrawAttr.PAINT_STROKE ? fillColor.hashCode() : 0);
	}

	public AttributeOption getPaintType() {
		return paintType;
	}

	public int getStrokeWidth() {
		return strokeWidth;
	}

	@Override
	@SuppressWarnings("unchecked")
	public <V> V getValue(Attribute<V> attr) {
		if (attr == DrawAttr.PAINT_TYPE)
			return (V) paintType;
		else if (attr == DrawAttr.STROKE_COLOR)
			return (V) strokeColor;
		else if (attr == DrawAttr.FILL_COLOR)
			return (V) fillColor;
		else if (attr == DrawAttr.STROKE_WIDTH)
			return (V) Integer.valueOf(strokeWidth);
		else
			return null;
	}

	@Override
	public void updateValue(Attribute<?> attr, Object value) {
		if (attr == DrawAttr.PAINT_TYPE) {
			paintType = (AttributeOption) value;
			fireAttributeListChanged();
		} else if (attr == DrawAttr.STROKE_COLOR)
			strokeColor = (Color) value;
		else if (attr == DrawAttr.FILL_COLOR)
			fillColor = (Color) value;
		else if (attr == DrawAttr.STROKE_WIDTH)
			strokeWidth = (Integer) value;
	}
}
